Business Intelligence Project    
Name: Zehra Ahmed, Farah Inayat, Kisa Fatima, Zuha Aqib    
Date: 18-May-2025

In [1]:
# print when the last code was run
from datetime import datetime
datetime.now().strftime("%Y-%m-%d %H:%M:%S")

'2025-05-27 17:55:53'

# Imports

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime, timedelta

### Some Functions
some preliminary functions we can use in the code

In [3]:
# Function to get current date and time as a string
def get_current_datetime():
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Data Loading

In [4]:
# Load the dataset
df = pd.read_csv('data/flights_sample_3m.csv')

pd.set_option('display.max_columns', None)

In [5]:
df

FL_DATE                 AIRLINE                 AIRLINE_DOT  \
0        2019-01-09   United Air Lines Inc.   United Air Lines Inc.: UA   
1        2022-11-19    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
2        2022-07-22   United Air Lines Inc.   United Air Lines Inc.: UA   
3        2023-03-06    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
4        2020-02-23        Spirit Air Lines        Spirit Air Lines: NK   
...             ...                     ...                         ...   
2999995  2022-11-13  American Airlines Inc.  American Airlines Inc.: AA   
2999996  2022-11-02  American Airlines Inc.  American Airlines Inc.: AA   
2999997  2022-09-11    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
2999998  2019-11-13        Republic Airline        Republic Airline: YX   
2999999  2019-06-15  Southwest Airlines Co.  Southwest Airlines Co.: WN   

        AIRLINE_CODE  DOT_CODE  FL_NUMBER ORIGIN          ORIGIN_CITY DEST  \
0                 UA     19977       1562    FLL  Fort Lauderdale, FL  EWR   
1                 DL     19790       1149    MSP      Minneapolis, MN  SEA   
2                 UA     19977        459    DEN           Denver, CO  MSP   
3                 DL     19790       2295    MSP      Minneapolis, MN  SFO   
4                 NK     20416        407    MCO          Orlando, FL  DFW   
...              ...       ...        ...    ...                  ...  ...   
2999995           AA     19805       1522    JAX     Jacksonville, FL  CLT   
2999996           AA     19805       1535    ORD          Chicago, IL  AUS   
2999997           DL     19790       2745    HSV       Huntsville, AL  ATL   
2999998           YX     20452       6134    BOS           Boston, MA  LGA   
2999999           WN     19393       2823    LGB       Long Beach, CA  SJC   

                     DEST_CITY  CRS_DEP_TIME  DEP_TIME  DEP_DELAY  TAXI_OUT  \
0                   Newark, NJ          1155    1151.0       -4.0      19.0   
1                  Seattle, WA          2120    2114.0       -6.0       9.0   
2              Minneapolis, MN           954    1000.0        6.0      20.0   
3            San Francisco, CA          1609    1608.0       -1.0      27.0   
4        Dallas/Fort Worth, TX          1840    1838.0       -2.0      15.0   
...                        ...           ...       ...        ...       ...   
2999995          Charlotte, NC          1742    1740.0       -2.0      10.0   
2999996             Austin, TX          1300    1254.0       -6.0      10.0   
2999997            Atlanta, GA           534     615.0       41.0      16.0   
2999998           New York, NY          1600    1555.0       -5.0      19.0   
2999999           San Jose, CA           730     727.0       -3.0       9.0   

         WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  ARR_TIME  ARR_DELAY  \
0            1210.0     1443.0      4.0          1501    1447.0      -14.0   
1            2123.0     2232.0     38.0          2315    2310.0       -5.0   
2            1020.0     1247.0      5.0          1252    1252.0        0.0   
3            1635.0     1844.0      9.0          1829    1853.0       24.0   
4            1853.0     2026.0     14.0          2041    2040.0       -1.0   
...             ...        ...      ...           ...       ...        ...   
2999995      1750.0     1845.0      6.0          1907    1851.0      -16.0   
2999996      1304.0     1514.0      5.0          1556    1519.0      -37.0   
2999997       631.0      759.0      6.0           729     805.0       36.0   
2999998      1614.0     1704.0      8.0          1728    1712.0      -16.0   
2999999       736.0      828.0      2.0           840     830.0      -10.0   

         CANCELLED CANCELLATION_CODE  DIVERTED  CRS_ELAPSED_TIME  \
0              0.0               NaN       0.0             186.0   
1              0.0               NaN       0.0             235.0   
2              0.0               NaN       0.0             118.0   
3              0

In [6]:
df.shape

(3000000, 32)

# Data Preparation

## Data Statistics

In [7]:
# Check for missing values
print("Missing values per column:")
df.isnull().sum()

Missing values per column:


FL_DATE                          0
AIRLINE                          0
AIRLINE_DOT                      0
AIRLINE_CODE                     0
DOT_CODE                         0
FL_NUMBER                        0
ORIGIN                           0
ORIGIN_CITY                      0
DEST                             0
DEST_CITY                        0
CRS_DEP_TIME                     0
DEP_TIME                     77615
DEP_DELAY                    77644
TAXI_OUT                     78806
WHEELS_OFF                   78806
WHEELS_ON                    79944
TAXI_IN                      79944
CRS_ARR_TIME                     0
ARR_TIME                     79942
ARR_DELAY                    86198
CANCELLED                        0
CANCELLATION_CODE          2920860
DIVERTED                         0
CRS_ELAPSED_TIME                14
ELAPSED_TIME                 86198
AIR_TIME                     86198
DISTANCE                         0
DELAY_DUE_CARRIER          2466137
DELAY_DUE_WEATHER   

In [8]:
# Display data types
print("\nData types:")
df.dtypes


Data types:


FL_DATE                     object
AIRLINE                     object
AIRLINE_DOT                 object
AIRLINE_CODE                object
DOT_CODE                     int64
FL_NUMBER                    int64
ORIGIN                      object
ORIGIN_CITY                 object
DEST                        object
DEST_CITY                   object
CRS_DEP_TIME                 int64
DEP_TIME                   float64
DEP_DELAY                  float64
TAXI_OUT                   float64
WHEELS_OFF                 float64
WHEELS_ON                  float64
TAXI_IN                    float64
CRS_ARR_TIME                 int64
ARR_TIME                   float64
ARR_DELAY                  float64
CANCELLED                  float64
CANCELLATION_CODE           object
DIVERTED                   float64
CRS_ELAPSED_TIME           float64
ELAPSED_TIME               float64
AIR_TIME                   float64
DISTANCE                   float64
DELAY_DUE_CARRIER          float64
DELAY_DUE_WEATHER   

In [9]:
# Get summary statistics
print("\nSummary statistics:")
df.describe(include='all')


Summary statistics:


FL_DATE                 AIRLINE                 AIRLINE_DOT  \
count      3000000                 3000000                     3000000   
unique        1704                      18                          18   
top     2019-07-25  Southwest Airlines Co.  Southwest Airlines Co.: WN   
freq          2379                  576470                      576470   
mean           NaN                     NaN                         NaN   
std            NaN                     NaN                         NaN   
min            NaN                     NaN                         NaN   
25%            NaN                     NaN                         NaN   
50%            NaN                     NaN                         NaN   
75%            NaN                     NaN                         NaN   
max            NaN                     NaN                         NaN   

       AIRLINE_CODE      DOT_CODE     FL_NUMBER   ORIGIN  ORIGIN_CITY  \
count       3000000  3.000000e+06  3.000000e+06  3000000      3000000   
unique           18           NaN           NaN      380          373   
top              WN           NaN           NaN      ATL  Chicago, IL   
freq         576470           NaN           NaN   153556       157368   
mean            NaN  1.997629e+04  2.511536e+03      NaN          NaN   
std             NaN  3.772846e+02  1.747258e+03      NaN          NaN   
min             NaN  1.939300e+04  1.000000e+00      NaN          NaN   
25%             NaN  1.979000e+04  1.051000e+03      NaN          NaN   
50%             NaN  1.993000e+04  2.152000e+03      NaN          NaN   
75%             NaN  2.036800e+04  3.797000e+03      NaN          NaN   
max             NaN  2.045200e+04  9.562000e+03      NaN          NaN   

           DEST    DEST_CITY  CRS_DEP_TIME      DEP_TIME     DEP_DELAY  \
count   3000000      3000000  3.000000e+06  2.922385e+06  2.922356e+06   
unique      380          373           NaN           NaN           NaN   
top         ATL  Chicago, IL           NaN           NaN           NaN   
freq     153569       158087           NaN           NaN           NaN   
mean        NaN          NaN  1.327062e+03  1.329776e+03  1.012333e+01   
std         NaN          NaN  4.858789e+02  4.993101e+02  4.925183e+01   
min         NaN          NaN  1.000000e+00  1.000000e+00 -9.000000e+01   
25%         NaN          NaN  9.150000e+02  9.160000e+02 -6.000000e+00   
50%         NaN          NaN  1.320000e+03  1.323000e+03 -2.000000e+00   
75%         NaN          NaN  1.730000e+03  1.739000e+03  6.000000e+00   
max         NaN          NaN  2.359000e+03  2.400000e+03  2.966000e+03   

            TAXI_OUT    WHEELS_OFF     WHEELS_ON       TAXI_IN  CRS_ARR_TIME  \
count   2.921194e+06  2.921194e+06  2.920056e+06  2.920056e+06  3.000000e+06   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    1.664305e+01  1.352361e+03  1.462500e+03  7.678982e+00  1.490561e+03   
std     9.192901e+00  5.008727e+02  5.272368e+02  6.269639e+00  5.115476e+02   
min     1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%     1.100000e+01  9.310000e+02  1.049000e+03  4.000000e+00  1.107000e+03   
50%     1.400000e+01  1.336000e+03  1.501000e+03  6.000000e+00  1.516000e+03   
75%     1.900000e+01  1.752000e+03  1.908000e+03  9.000000e+00  1.919000e+03   
max     1.840000e+02  2.400000e+03  2.400000e+03  2.490000e+02  2.400000e+03   

            ARR_TIME     ARR_DELAY     CANCELLED CANCELLATION_CODE  \
count   2.920058e+06  2.913802e+06  3.000000e+06             79140   
unique           NaN           NaN           NaN                 4   
top              NaN           NaN           NaN                 B   
freq             NaN           NaN           NaN             28772   
mean    1.466511e+03  4.260858e+00  2.638000e

## Changing Data Type

### Defined Functions
here are some functions we can use for changing the datatype

In [10]:
def to_string(df, column_name):
    """
    Convert a pandas DataFrame column to a string.
    """
    df[column_name] = df[column_name].astype('string')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [11]:
def to_date(df, column_name, format=None):
    """
    Convert a pandas DataFrame column to datetime.
    """
    df[column_name] = pd.to_datetime(df[column_name], format=format)
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df

In [12]:
def to_int(df, column_name):
    """
    Convert a pandas DataFrame column to a int.
    """
    df[column_name] = df[column_name].astype('int64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [13]:
def to_float(df, column_name):
    """
    Convert a pandas DataFrame column to a float.
    """
    df[column_name] = df[column_name].astype('float64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

### Implementation
so there is a date column lets make it a type date column, and there is a few columns which are object datatype, lets change to string for better readability

In [14]:
df = to_date(df, 'FL_DATE', '%Y-%m-%d')

Column Name: 'FL_DATE', Type: datetime64[ns]


In [15]:
df = to_string(df, 'AIRLINE')
df = to_string(df, 'AIRLINE_DOT')
df = to_string(df, 'AIRLINE_CODE')
df = to_string(df, 'ORIGIN')
df = to_string(df, 'ORIGIN_CITY')
df = to_string(df, 'DEST')
df = to_string(df, 'DEST_CITY')
df = to_string(df, 'CANCELLATION_CODE')

Column Name: 'AIRLINE', Type: string
Column Name: 'AIRLINE_DOT', Type: string
Column Name: 'AIRLINE_CODE', Type: string
Column Name: 'ORIGIN', Type: string
Column Name: 'ORIGIN_CITY', Type: string
Column Name: 'DEST', Type: string
Column Name: 'DEST_CITY', Type: string
Column Name: 'CANCELLATION_CODE', Type: string


## Fixing columns saath saath
so for each column we're just gonna check is it coherenet? are there any errors like mistypes? this is for only those columns which are text and have some textual information and their unique values are small and checkable

### AIRLINE
here lets extract its 18 unique values and check it

In [16]:
# Just checking unique airline names
print(df['AIRLINE'].value_counts())

AIRLINE
Southwest Airlines Co.                576470
Delta Air Lines Inc.                  395239
American Airlines Inc.                383106
SkyWest Airlines Inc.                 343737
United Air Lines Inc.                 254504
Republic Airline                      143107
Envoy Air                             121256
JetBlue Airways                       112844
Endeavor Air Inc.                     112463
PSA Airlines Inc.                     107050
Alaska Airlines Inc.                  100467
Spirit Air Lines                       95711
Mesa Airlines Inc.                     65012
Frontier Airlines Inc.                 64466
Allegiant Air                          52738
Hawaiian Airlines Inc.                 32114
Horizon Air                            20634
ExpressJet Airlines LLC d/b/a aha!     19082
Name: count, dtype: Int64


so there is no typo. lets move ahead to the next one

### AIRLINE DOT

In [17]:
print(df['AIRLINE_DOT'].value_counts())

AIRLINE_DOT
Southwest Airlines Co.: WN                576470
Delta Air Lines Inc.: DL                  395239
American Airlines Inc.: AA                383106
SkyWest Airlines Inc.: OO                 343737
United Air Lines Inc.: UA                 254504
Republic Airline: YX                      143107
Envoy Air: MQ                             121256
JetBlue Airways: B6                       112844
Endeavor Air Inc.: 9E                     112463
PSA Airlines Inc.: OH                     107050
Alaska Airlines Inc.: AS                  100467
Spirit Air Lines: NK                       95711
Mesa Airlines Inc.: YV                     65012
Frontier Airlines Inc.: F9                 64466
Allegiant Air: G4                          52738
Hawaiian Airlines Inc.: HA                 32114
Horizon Air: QX                            20634
ExpressJet Airlines LLC d/b/a aha!: EV     19082
Name: count, dtype: Int64


nothing again

### AIRLINE CODE

In [18]:
print(df['AIRLINE_CODE'].value_counts())

AIRLINE_CODE
WN    576470
DL    395239
AA    383106
OO    343737
UA    254504
YX    143107
MQ    121256
B6    112844
9E    112463
OH    107050
AS    100467
NK     95711
YV     65012
F9     64466
G4     52738
HA     32114
QX     20634
EV     19082
Name: count, dtype: Int64


nothing again but the column `AIRLINE_DOT` is very similar to the columns we already have - its just a concatentation of `AIRLINES` AND `AIRLINE_CODE`. So lets drop it.

In [19]:
df = df.drop('AIRLINE_DOT', axis=1)

# Missing Values
here we handle missing values for each column. 

## DEP_TIME

In [20]:
df['DEP_TIME'].value_counts()

DEP_TIME
555.0    7991
557.0    7186
556.0    7184
655.0    6662
558.0    6640
         ... 
406.0       4
349.0       4
408.0       3
434.0       3
415.0       3
Name: count, Length: 1440, dtype: int64

In [21]:
df['DEP_TIME'].isnull().sum()

77615

so we see that all the missing values are either 77615 or more. we have an intuition that maybe those rows are those rows that have ALL the missing values. so we get all the missing values which are 77615 >= and extract their rows and see how many are extracted. if they are 77615, that means all those rows have majority columns empty - they have the important columns empty

In [22]:
null_rows = df[
    df['DEP_TIME'].isnull() & 
    df['DEP_DELAY'].isnull() & 
    df['TAXI_OUT'].isnull() & 
    df['WHEELS_OFF'].isnull() & 
    df['WHEELS_ON'].isnull() & 
    df['TAXI_IN'].isnull() & 
    df['ARR_TIME'].isnull() & 
    df['ARR_DELAY'].isnull() & 
    df['ELAPSED_TIME'].isnull() &
    df['AIR_TIME'].isnull() &
    df['DELAY_DUE_CARRIER'].isnull() &
    df['DELAY_DUE_WEATHER'].isnull() &
    df['DELAY_DUE_NAS'].isnull() &
    df['DELAY_DUE_SECURITY'].isnull() &
    df['DELAY_DUE_LATE_AIRCRAFT'].isnull()
]

null_rows

FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
14      2020-04-07  Southwest Airlines Co.           WN     19393        687   
45      2022-01-02        Republic Airline           YX     20452       3467   
80      2020-04-20  Southwest Airlines Co.           WN     19393        246   
191     2021-12-31    Alaska Airlines Inc.           AS     19930       1097   
236     2020-04-14  Southwest Airlines Co.           WN     19393        446   
...            ...                     ...          ...       ...        ...   
2999779 2019-06-17       PSA Airlines Inc.           OH     20397       5305   
2999865 2021-10-22   SkyWest Airlines Inc.           OO     20304       3260   
2999910 2020-04-08           Allegiant Air           G4     20368       1595   
2999943 2020-03-30  American Airlines Inc.           AA     19805       1877   
2999946 2023-07-09        Republic Airline           YX     20452       4296   

        ORIGIN          ORIGIN_CITY DEST        DEST_CITY  CRS_DEP_TIME  \
14         SJC         San Jose, CA  LAX  Los Angeles, CA          2155   
45         ORD          Chicago, IL  DCA   Washington, DC          1935   
80         BNA        Nashville, TN  ATL      Atlanta, GA          1055   
191        IAD       Washington, DC  LAX  Los Angeles, CA          1720   
236        LAS        Las Vegas, NV  HOU      Houston, TX           500   
...        ...                  ...  ...              ...           ...   
2999779    DCA       Washington, DC  CAK        Akron, OH          1959   
2999865    ORD          Chicago, IL  MEM      Memphis, TN          1530   
2999910    PGD      Punta Gorda, FL  STC    St. Cloud, MN          1504   
2999943    FLL  Fort Lauderdale, FL  CLT    Charlotte, NC          1443   
2999946    BOS           Boston, MA  JFK     New York, NY          1734   

         DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  \
14            NaN        NaN       NaN         NaN        NaN      NaN   
45            NaN        NaN       NaN         NaN        NaN      NaN   
80            NaN        NaN       NaN         NaN        NaN      NaN   
191           NaN        NaN       NaN         NaN        NaN      NaN   
236           NaN        NaN       NaN         NaN        NaN      NaN   
...           ...        ...       ...         ...        ...      ...   
2999779       NaN        NaN       NaN         NaN        NaN      NaN   
2999865       NaN        NaN       NaN         NaN        NaN      NaN   
2999910       NaN        NaN       NaN         NaN        NaN      NaN   
2999943       NaN        NaN       NaN         NaN        NaN      NaN   
2999946       NaN        NaN       NaN         NaN        NaN      NaN   

         CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
14               2315       NaN        NaN        1.0                 D   
45               2241       NaN        NaN        1.0                 B   
80               1300       NaN        NaN        1.0                 D   
191              2024       NaN        NaN        1.0                 A   
236               955       NaN        NaN        1.0                 D   
...               ...       ...        ...        ...               ...   
2999779          2112       NaN        NaN        1.0                 C   
2999865          1719       NaN        NaN        1.0                 A   
2999910          1742       NaN        NaN        1.0                 D   
2999943          1655       NaN        NaN        1.0                 D   
2999946          1855       NaN        NaN        1.0                 C   

         DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
14            0.0              80.0           NaN       NaN     308.0   
45            0.0             126.0           NaN       NaN     612.0   
80            0.0              65.0           NaN       NaN     214.0   
191           0.0             364.0           NaN       NaN    2288

so we see here that those 77615 rows have majority columns empty. so because we don't have the DEP_time and DEP_delay then its better to just drop them. there is not ELAPSED_TIME, ARR_TIME then filling with median/mean values would ruin our analysis - better to drop.

In [23]:
df = df.drop(null_rows.index)

In [24]:
df.isnull().sum()

FL_DATE                          0
AIRLINE                          0
AIRLINE_CODE                     0
DOT_CODE                         0
FL_NUMBER                        0
ORIGIN                           0
ORIGIN_CITY                      0
DEST                             0
DEST_CITY                        0
CRS_DEP_TIME                     0
DEP_TIME                         0
DEP_DELAY                       29
TAXI_OUT                      1191
WHEELS_OFF                    1191
WHEELS_ON                     2329
TAXI_IN                       2329
CRS_ARR_TIME                     0
ARR_TIME                      2327
ARR_DELAY                     8583
CANCELLED                        0
CANCELLATION_CODE          2920860
DIVERTED                         0
CRS_ELAPSED_TIME                 1
ELAPSED_TIME                  8583
AIR_TIME                      8583
DISTANCE                         0
DELAY_DUE_CARRIER          2388522
DELAY_DUE_WEATHER          2388522
DELAY_DUE_NAS       

## DEP_DELAY

okay so we see that CRS_DEP_TIME and DEP_TIME is 0 null values that means we can compute DEP_DELAY null values using DEP_TIME - CRS_DEP_TIME to find the delay. however as both are in hhmm we will need to convert to minutes first 

In [25]:
def convert_to_minutes(t):
    if pd.isnull(t):
        return np.nan
    t = int(t)
    hours = t // 100
    minutes = t % 100
    return hours * 60 + minutes

## TAXI_OUT

## WHEELS_OFF

## WHEELS_ON

## TAXI_IN

## ARR_TIME

## ARR_DELAY

## CANCELLATION_CODE
so we can see that the cancellation_code empty values are BECAUSE THAT FLIGHT WAS NOT CANCELLED. AS IT IS A STRING type we can just fill with a string of "not cancelled"

In [26]:
df['CANCELLATION_CODE'].value_counts()

CANCELLATION_CODE
B    762
A    499
C    256
D      8
Name: count, dtype: Int64

so here we see that this is not interpretable. we searched up and dug up the internet to get to know that from the U.S. Department of Transportation (DOT) flight data standards, here is what each cancellation code means:
- A = Carrier — airline-related (crew, aircraft, etc.)
- B = Weather — adverse weather at origin/destination or en route
- C = NAS — National Airspace System (ATC, delays, airport congestion)
- D = Security — e.g., TSA delays, threats, evacuations     

so instead of leaving it just like this, lets make it more visualizeable.

along with mapping each one, lets replace all null values with "Not Cancelled"

In [27]:
cancel_map = {
    'A': 'Carrier',
    'B': 'Weather',
    'C': 'Airspace System (NAS)',
    'D': 'Security',
    'Not Cancelled': 'Not Cancelled'
}

df['CANCELLATION_CODE'] = df['CANCELLATION_CODE'].fillna('Not Cancelled').map(cancel_map)

In [28]:
df['CANCELLATION_CODE'].value_counts()

CANCELLATION_CODE
Not Cancelled            2920860
Weather                      762
Carrier                      499
Airspace System (NAS)        256
Security                       8
Name: count, dtype: int64

## CRS_ELAPSED_TIME

## ELAPSED_TIME

## AIR_TIME

## DELAY_DUE_CARRIER, DELAY_DUE_WEATHER, DELAY_DUE_NAS, DELAY_DUE_SECURITY, DELAY_DUE_LATE_AIRCRAFT
now every column here has one thing in common: same missing values. this also means that whenever there was NOT A DELAY, then they were left blank. so it is better to fill them with 0s to show that there was NO DELAY. if i fill them with -1, it will hinder the mean/median/average, and saying -1minutes of delay does not make sense, wheras 0 minutes of delay does.

In [29]:
df['DELAY_DUE_CARRIER'].value_counts()

DELAY_DUE_CARRIER
0.0       236912
15.0        9102
1.0         9073
2.0         8941
3.0         8670
           ...  
1117.0         1
2565.0         1
998.0          1
1192.0         1
1859.0         1
Name: count, Length: 1291, dtype: int64

so we see from the value_counts that the majority is 0s which means there is no delay of this cause. so we will fill all NANs wth 0s

In [30]:
delay_columns = ['DELAY_DUE_CARRIER', 'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY', 'DELAY_DUE_LATE_AIRCRAFT']
for col in delay_columns:
    df[col] = df[col].fillna(0.0)
    
df.isnull().sum()

FL_DATE                       0
AIRLINE                       0
AIRLINE_CODE                  0
DOT_CODE                      0
FL_NUMBER                     0
ORIGIN                        0
ORIGIN_CITY                   0
DEST                          0
DEST_CITY                     0
CRS_DEP_TIME                  0
DEP_TIME                      0
DEP_DELAY                    29
TAXI_OUT                   1191
WHEELS_OFF                 1191
WHEELS_ON                  2329
TAXI_IN                    2329
CRS_ARR_TIME                  0
ARR_TIME                   2327
ARR_DELAY                  8583
CANCELLED                     0
CANCELLATION_CODE             0
DIVERTED                      0
CRS_ELAPSED_TIME              1
ELAPSED_TIME               8583
AIR_TIME                   8583
DISTANCE                      0
DELAY_DUE_CARRIER             0
DELAY_DUE_WEATHER             0
DELAY_DUE_NAS                 0
DELAY_DUE_SECURITY            0
DELAY_DUE_LATE_AIRCRAFT       0
dtype: i

## New columns
here we analyse if we need new columns in integers or not for example of destination city etc so that we can see common ones

### FL_DATE
here we add 4 new columns for year, quarter, month, day to make visualisation easier

In [31]:
# for FL_DATE, add 4 new columns which have the year, quarter, month, date for better preprocessing and display
df['year'] = df['FL_DATE'].dt.year
df['quarter'] = df['FL_DATE'].dt.quarter
df['month'] = df['FL_DATE'].dt.strftime('%b')  # 3-letter month name
df = to_string(df, 'month')
df['day'] = df['FL_DATE'].dt.day

df

Column Name: 'month', Type: string


FL_DATE                 AIRLINE AIRLINE_CODE  DOT_CODE  FL_NUMBER  \
0       2019-01-09   United Air Lines Inc.           UA     19977       1562   
1       2022-11-19    Delta Air Lines Inc.           DL     19790       1149   
2       2022-07-22   United Air Lines Inc.           UA     19977        459   
3       2023-03-06    Delta Air Lines Inc.           DL     19790       2295   
4       2020-02-23        Spirit Air Lines           NK     20416        407   
...            ...                     ...          ...       ...        ...   
2999995 2022-11-13  American Airlines Inc.           AA     19805       1522   
2999996 2022-11-02  American Airlines Inc.           AA     19805       1535   
2999997 2022-09-11    Delta Air Lines Inc.           DL     19790       2745   
2999998 2019-11-13        Republic Airline           YX     20452       6134   
2999999 2019-06-15  Southwest Airlines Co.           WN     19393       2823   

        ORIGIN          ORIGIN_CITY DEST              DEST_CITY  CRS_DEP_TIME  \
0          FLL  Fort Lauderdale, FL  EWR             Newark, NJ          1155   
1          MSP      Minneapolis, MN  SEA            Seattle, WA          2120   
2          DEN           Denver, CO  MSP        Minneapolis, MN           954   
3          MSP      Minneapolis, MN  SFO      San Francisco, CA          1609   
4          MCO          Orlando, FL  DFW  Dallas/Fort Worth, TX          1840   
...        ...                  ...  ...                    ...           ...   
2999995    JAX     Jacksonville, FL  CLT          Charlotte, NC          1742   
2999996    ORD          Chicago, IL  AUS             Austin, TX          1300   
2999997    HSV       Huntsville, AL  ATL            Atlanta, GA           534   
2999998    BOS           Boston, MA  LGA           New York, NY          1600   
2999999    LGB       Long Beach, CA  SJC           San Jose, CA           730   

         DEP_TIME  DEP_DELAY  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  \
0          1151.0       -4.0      19.0      1210.0     1443.0      4.0   
1          2114.0       -6.0       9.0      2123.0     2232.0     38.0   
2          1000.0        6.0      20.0      1020.0     1247.0      5.0   
3          1608.0       -1.0      27.0      1635.0     1844.0      9.0   
4          1838.0       -2.0      15.0      1853.0     2026.0     14.0   
...           ...        ...       ...         ...        ...      ...   
2999995    1740.0       -2.0      10.0      1750.0     1845.0      6.0   
2999996    1254.0       -6.0      10.0      1304.0     1514.0      5.0   
2999997     615.0       41.0      16.0       631.0      759.0      6.0   
2999998    1555.0       -5.0      19.0      1614.0     1704.0      8.0   
2999999     727.0       -3.0       9.0       736.0      828.0      2.0   

         CRS_ARR_TIME  ARR_TIME  ARR_DELAY  CANCELLED CANCELLATION_CODE  \
0                1501    1447.0      -14.0        0.0     Not Cancelled   
1                2315    2310.0       -5.0        0.0     Not Cancelled   
2                1252    1252.0        0.0        0.0     Not Cancelled   
3                1829    1853.0       24.0        0.0     Not Cancelled   
4                2041    2040.0       -1.0        0.0     Not Cancelled   
...               ...       ...        ...        ...               ...   
2999995          1907    1851.0      -16.0        0.0     Not Cancelled   
2999996          1556    1519.0      -37.0        0.0     Not Cancelled   
2999997           729     805.0       36.0        0.0     Not Cancelled   
2999998          1728    1712.0      -16.0        0.0     Not Cancelled   
2999999           840     830.0      -10.0        0.0     Not Cancelled   

         DIVERTED  CRS_ELAPSED_TIME  ELAPSED_TIME  AIR_TIME  DISTANCE  \
0             0.0             186.0         176.0     153.0    1065.0   
1             0.0             235.0         236.0     189.0    1399.0   
2             0.0             118.0         112.0      87.0     680.

In [32]:
# Move the new columns after FL_DATE
# Get all column names
cols = list(df.columns)

# Get index of FL_DATE
fl_date_index = cols.index('FL_DATE')

# Remove the new columns from the end
for col in ['year', 'quarter', 'month', 'day']:
    cols.remove(col)

# Insert new columns right after FL_DATE
for i, col in enumerate(['year', 'quarter', 'month', 'day']):
    cols.insert(fl_date_index + 1 + i, col)

# Reorder the DataFrame
df = df[cols]

### Time columns
all the time columns are just integer, floats like 1210, 2123, 1020 etc which basically means 12:10, 21:23, 10:20 etc. so we add new columns which writes in date format for easier understanding

In [33]:
def convert_hhmm(val):
    try:
        val = int(val)
        hour = val // 100
        minute = val % 100
        if 0 <= hour < 24 and 0 <= minute < 60:
            return f"{hour:02d}:{minute:02d}"
    except:
        pass
    return None

In [34]:
time_columns = ['CRS_DEP_TIME', 'DEP_TIME', 'WHEELS_OFF', 'WHEELS_ON', 'CRS_ARR_TIME', 'ARR_TIME']

for col in time_columns:
    df[col.lower() + '_converted'] = df[col].apply(convert_hhmm)
    df = to_string(df, col.lower() + '_converted')
    cols = list(df.columns)
    index = cols.index(col)
    cols.remove(col.lower() + '_converted')
    cols.insert(index + 1, col.lower() + '_converted')
    df = df[cols]

C:\Users\DELL\AppData\Local\Temp\ipykernel_21088\2966946087.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col.lower() + '_converted'] = df[col].apply(convert_hhmm)
C:\Users\DELL\AppData\Local\Temp\ipykernel_21088\70885104.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column_name] = df[column_name].astype('string')


Column Name: 'crs_dep_time_converted', Type: string
Column Name: 'dep_time_converted', Type: string
Column Name: 'wheels_off_converted', Type: string
Column Name: 'wheels_on_converted', Type: string
Column Name: 'crs_arr_time_converted', Type: string
Column Name: 'arr_time_converted', Type: string


### Categorical Columns
so the CANCELLED column and DIVERTED column are according to `dictionary.html` binary values of 0 1, so to improve visualisation in BI, we can convert them to categorical values of 'No' and 'Yes' respectively.

In [35]:
df['CANCELLED'].value_counts()

CANCELLED
0.0    2920860
1.0       1525
Name: count, dtype: int64

In [36]:
df['DIVERTED'].value_counts()

DIVERTED
0.0    2915329
1.0       7056
Name: count, dtype: int64

so we can convert CANCELLED and DIVERTED to integers, and then add columns for categorical

In [37]:
df = to_int(df, 'CANCELLED')
df = to_int(df, 'DIVERTED')

Column Name: 'CANCELLED', Type: int64
Column Name: 'DIVERTED', Type: int64


In [38]:
df['DIVERTED'].value_counts()

DIVERTED
0    2915329
1       7056
Name: count, dtype: int64

In [39]:
df['CANCELLED'].value_counts()

CANCELLED
0    2920860
1       1525
Name: count, dtype: int64

In [40]:
# Mapping for a new column
mapping = {
    0: 'No', 
    1: 'Yes'
}

# Create a new column 
df['cancelled_c'] = df['CANCELLED'].map(mapping)
df = to_string(df, 'cancelled_c')

df['diverted_c'] = df['DIVERTED'].map(mapping)
df = to_string(df, 'diverted_c')

Column Name: 'cancelled_c', Type: string
Column Name: 'diverted_c', Type: string


In [41]:
df['cancelled_c'].value_counts()

cancelled_c
No     2920860
Yes       1525
Name: count, dtype: Int64

In [42]:
df['diverted_c'].value_counts()

diverted_c
No     2915329
Yes       7056
Name: count, dtype: Int64

## Changing Column names
to improve understandability

In [43]:
df.rename(columns={'CRS_DEP_TIME': 'SCHEDULED_DEP_TIME'}, inplace=True)
df.rename(columns={'CRS_ARR_TIME': 'SCHEDULED_ARR_TIME'}, inplace=True)
df.rename(columns={'CRS_ELAPSED_TIME': 'SCHEDULED_ELAPSED_TIME'}, inplace=True)
df.rename(columns={'FL_DATE': 'FLIGHT_DATE'}, inplace=True)
df.rename(columns={'FL_NUMBER': 'FLIGHT_NUMBER'}, inplace=True)

# Recheck
a simple recheck of all data statistics to see if the data is fine before donwloading

In [44]:
# Check for missing values
print("Missing values per column:")
df.isnull().sum()

Missing values per column:


FLIGHT_DATE                   0
year                          0
quarter                       0
month                         0
day                           0
AIRLINE                       0
AIRLINE_CODE                  0
DOT_CODE                      0
FLIGHT_NUMBER                 0
ORIGIN                        0
ORIGIN_CITY                   0
DEST                          0
DEST_CITY                     0
SCHEDULED_DEP_TIME            0
crs_dep_time_converted        0
DEP_TIME                      0
dep_time_converted          242
DEP_DELAY                    29
TAXI_OUT                   1191
WHEELS_OFF                 1191
wheels_off_converted       1592
WHEELS_ON                  2329
wheels_on_converted        3438
TAXI_IN                    2329
SCHEDULED_ARR_TIME            0
crs_arr_time_converted       13
ARR_TIME                   2327
arr_time_converted         3743
ARR_DELAY                  8583
CANCELLED                     0
CANCELLATION_CODE             0
DIVERTED

In [45]:
# Display data types
print("\nData types:")
df.dtypes


Data types:


FLIGHT_DATE                datetime64[ns]
year                                int32
quarter                             int32
month                      string[python]
day                                 int32
AIRLINE                    string[python]
AIRLINE_CODE               string[python]
DOT_CODE                            int64
FLIGHT_NUMBER                       int64
ORIGIN                     string[python]
ORIGIN_CITY                string[python]
DEST                       string[python]
DEST_CITY                  string[python]
SCHEDULED_DEP_TIME                  int64
crs_dep_time_converted     string[python]
DEP_TIME                          float64
dep_time_converted         string[python]
DEP_DELAY                         float64
TAXI_OUT                          float64
WHEELS_OFF                        float64
wheels_off_converted       string[python]
WHEELS_ON                         float64
wheels_on_converted        string[python]
TAXI_IN                           

In [46]:
df

FLIGHT_DATE  year  quarter month  day                 AIRLINE  \
0        2019-01-09  2019        1   Jan    9   United Air Lines Inc.   
1        2022-11-19  2022        4   Nov   19    Delta Air Lines Inc.   
2        2022-07-22  2022        3   Jul   22   United Air Lines Inc.   
3        2023-03-06  2023        1   Mar    6    Delta Air Lines Inc.   
4        2020-02-23  2020        1   Feb   23        Spirit Air Lines   
...             ...   ...      ...   ...  ...                     ...   
2999995  2022-11-13  2022        4   Nov   13  American Airlines Inc.   
2999996  2022-11-02  2022        4   Nov    2  American Airlines Inc.   
2999997  2022-09-11  2022        3   Sep   11    Delta Air Lines Inc.   
2999998  2019-11-13  2019        4   Nov   13        Republic Airline   
2999999  2019-06-15  2019        2   Jun   15  Southwest Airlines Co.   

        AIRLINE_CODE  DOT_CODE  FLIGHT_NUMBER ORIGIN          ORIGIN_CITY  \
0                 UA     19977           1562    FLL  Fort Lauderdale, FL   
1                 DL     19790           1149    MSP      Minneapolis, MN   
2                 UA     19977            459    DEN           Denver, CO   
3                 DL     19790           2295    MSP      Minneapolis, MN   
4                 NK     20416            407    MCO          Orlando, FL   
...              ...       ...            ...    ...                  ...   
2999995           AA     19805           1522    JAX     Jacksonville, FL   
2999996           AA     19805           1535    ORD          Chicago, IL   
2999997           DL     19790           2745    HSV       Huntsville, AL   
2999998           YX     20452           6134    BOS           Boston, MA   
2999999           WN     19393           2823    LGB       Long Beach, CA   

        DEST              DEST_CITY  SCHEDULED_DEP_TIME  \
0        EWR             Newark, NJ                1155   
1        SEA            Seattle, WA                2120   
2        MSP        Minneapolis, MN                 954   
3        SFO      San Francisco, CA                1609   
4        DFW  Dallas/Fort Worth, TX                1840   
...      ...                    ...                 ...   
2999995  CLT          Charlotte, NC                1742   
2999996  AUS             Austin, TX                1300   
2999997  ATL            Atlanta, GA                 534   
2999998  LGA           New York, NY                1600   
2999999  SJC           San Jose, CA                 730   

        crs_dep_time_converted  DEP_TIME dep_time_converted  DEP_DELAY  \
0                        11:55    1151.0              11:51       -4.0   
1                        21:20    2114.0              21:14       -6.0   
2                        09:54    1000.0              10:00        6.0   
3                        16:09    1608.0              16:08       -1.0   
4                        18:40    1838.0              18:38       -2.0   
...                        ...       ...                ...        ...   
2999995                  17:42    1740.0              17:40       -2.0   
2999996                  13:00    1254.0              12:54       -6.0   
2999997                  05:34     615.0              06:15       41.0   
2999998                  16:00    1555.0              15:55       -5.0   
2999999                  07:30     727.0              07:27       -3.0   

         TAXI_OUT  WHEELS_OFF wheels_off_converted  WHEELS_ON  \
0            19.0      1210.0                12:10     1443.0   
1             9.0      2123.0                21:23     2232.0   
2            20.0      1020.0                10:20     1247.0   
3            27.0      1635.0                16:35     1844.0   
4            15.0      1853.0                18:53     2026.0   
...           ...         ...                  ...        ...   
2999995      10.0      1750.0                17:50     1845.0   
2999996      10.0      1304.0                13:04     1514.0   
2999997      16.0      

# Download cleaned dataset

In [47]:
# Save the cleaned DataFrame to a CSV file
# df.to_csv('data/flights_sample_3m_cleaned.csv', index=False)

In [48]:
# Extract first 1000 rows
df_1000 = df.head(1000)

# Save to CSV with column headers
df_1000.to_csv('data/flights_sample_1k_cleaned.csv', index=False)

In [49]:
# revert pandas settings back to normal
# pd.reset_option('display.max_columns')

# Last run

In [50]:
get_current_datetime()

'2025-05-27 17:57:44'